pkg/orchestrator/controller/controller.go

```diff
@@ -0,0 +1,266 @@
+	label, exists = labels[key]
+	return
+}
+
+func setLabel(obj runtime.Object, key, value string) runtime.Object {
+	copy := obj.DeepCopyObject()
+
+	// We can safely ignore the error as all the resources implement meta.Object
+	metadata, _ := meta.Accessor(copy)
+
+	labels := metadata.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+
+	labels[key] = value
+	metadata.SetLabels(labels)
+	return copy
+}
+
+func newResourceEventHandler(clusterID string, workQueue workqueue.RateLimitingInterface) cache.ResourceEventHandler {
+	enqueue := func(obj interface{}, keyFunc func(obj interface{}) (string, error)) {
+		key, err := keyFunc(obj)
+		if err != nil {
+			klog.Errorf("Error creating MetaNamespaceKey: %v", err)
+			return
+		}
+
+		if label, exists := getLabel(obj, clusterIDLabelKey); exists && label != clusterID {
+			klog.V(2).Infof("Label \"%s=%s\" for \"%s\" does not match \"%s\" - skipping", clusterIDLabelKey,
+				label, key, clusterID)
+			return
+		}
+
+		workQueue.Add(key)
+	}
+
+	return &cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			enqueue(obj, cache.MetaNamespaceKeyFunc)
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			// TODO - handle updates or are the resources immutable?
+		},
+		DeleteFunc: func(obj interface{}) {
+			enqueue(obj, cache.DeletionHandlingMetaNamespaceKeyFunc)
+		},
+	}
+}
+
+func (c *controller) OnAdd(clusterID string, kubeConfig *rest.Config) {
+	klog.V(2).Infof("In controller OnAdd for cluster %s", clusterID)
+
+	c.startNewWatcherForCluster(kubeConfig, clusterID)
+
+	klog.Infof("Cluster \"%s\" has been added - started submariner informers", clusterID)
+}
+
+func (c *controller) OnUpdate(clusterID string, kubeConfig *rest.Config) {
+	klog.V(2).Infof("In controller OnUpdate for cluster %s", clusterID)
+
+	c.startNewWatcherForCluster(kubeConfig, clusterID)
+
+	klog.Infof("Cluster \"%s\" has been updated - restarted submariner informers", clusterID)
+}
+
+func (c *controller) OnRemove(clusterID string) {
+	klog.V(2).Infof("In controller OnRemove for cluster %s", clusterID)
+
+	c.Lock()
+	defer c.Unlock()
+
+	clusterWatcher := c.clusterWatcherMap[clusterID]
+	if clusterWatcher == nil {
+		klog.V(2).Infof("clusterWatcher entry for cluster %s not found", clusterID)
+		return
+	}
+
+	clusterWatcher.close()
+	delete(c.clusterWatcherMap, clusterID)
+
+	klog.Infof("Cluster %s has been removed - stopped submariner informers", clusterID)
+}
+
+func runQueueWorker(clusterID string, federator federate.Federator, informer cache.SharedIndexInformer,
+	workQueue workqueue.RateLimitingInterface) {
+
+	for {
+		key, shutdown := workQueue.Get()
+		if shutdown {
+			klog.V(2).Infof("Submariner watcher for cluster \"%s\" stopped", clusterID)
+			return
+		}
+
+		klog.V(2).Infof("Processing key \"%s\" for cluster \"%s\"", key, clusterID)
+
+		func() {
+			defer workQueue.Done(key)
+
+			item, exists, err := informer.GetIndexer().GetByKey(key.(string))
+			if err != nil {
```

Comment:
yeah but I kind of preferred to use `if-elif-else` in this case since the conditions are related and it looks more natural to me. Also using the `return` approach actually adds more lines and I don't think it simplifies it.

Reply:
I actually tend to agree with the "return" method since it'll reduce the level of indentation which is already quite high (3 levels in for the normal flow), and make the code more readable since the normal flow is all on the same indentation level, while the exceptional cases are the ones that stray from the normal path.

Reply:
sure - I'm flexible either way - can refactor in followup.